### PR TITLE
🌱Add more helper functions in topologymutation varaible.go to help unmarshal variables

### PR DESCRIPTION
**Pull request by lubronzhan**
**What this PR does / why we need it**:
Add more function. For example to unmarshal nested variable from topology spec, and unmarshal bool variable from topology spec.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes https://github.com/kubernetes-sigs/cluster-api/issues/9669

### CODE REVIEW

File: exp/runtime/topologymutation/variables.go [hunk 1]
```diff
@@ -53,3 +56,46 @@ func GetStringVariable(templateVariables map[string]apiextensionsv1.JSON, variab
 	}
 	return stringValue, true, nil
 }
+
+// GetBoolVariable get the value as a bool.
+func GetBoolVariable(templateVariables map[string]apiextensionsv1.JSON, variableName string) (bool, bool, error) {
```
Comment (killianmuldoon) [inline exp/runtime/topologymutation/variables.go:61]: Not sure it should be a blocker here but `bool, book, error` as a return value seems pretty confusing and easy to get wrong.

I think we should either have a custom result type, or else a custom error which will allow us to check for the `variableNotFound` case. I think the custom error is probably the better option and matches what we do elsewhere.

We should adopt the same pattern for the other `Get*Variable` functions.

File: exp/runtime/topologymutation/variables.go [hunk 2]
```diff
@@ -53,3 +56,46 @@ func GetStringVariable(templateVariables map[string]apiextensionsv1.JSON, variab
...
+	boolValue, err := strconv.ParseBool(stringValue)
+	if err != nil {
+		return false, true, err
+	}
+	return boolValue, true, nil
+}
+
+// GetVariableObject returns get variable's string value then unmarshal it into
+// object passed from v.
+func GetVariableObject(templateVariables map[string]apiextensionsv1.JSON, variableName string, v interface{}) (bool, error) {
```
Comment (killianmuldoon) [inline exp/runtime/topologymutation/variables.go:82]: I think we should to change the name of this function to differentiate it from the `Bool` and `String` functions which return the variable. 

Maybe `GetObjectVariableInto`? and change the `v` to `into`

File: exp/runtime/topologymutation/variables_test.go [hunk 3]
```diff
@@ -130,3 +131,147 @@ func Test_GetStringTemplateVariable(t *testing.T) {
...
+			expectedValue:           false,
+			expectedFound:           false,
+			object:                  Network{},
+			expepctedVariableObject: Network{},
+			expectedErr:             true,
+		},
+		{
+			name:                    "variable not found",
+			variables:               nil,
+			variableName:            "notEsists",
```
Comment (killianmuldoon) [inline exp/runtime/topologymutation/variables_test.go:224]: nit

```suggestion
			variableName:            "notExists",
```

File: exp/runtime/topologymutation/variables_test.go [hunk 4]
```diff
@@ -130,3 +131,147 @@ func Test_GetStringTemplateVariable(t *testing.T) {
...
+			variables: map[string]apiextensionsv1.JSON{
+				"node":    {Raw: []byte(`{"name": "a"}`)},
+				"network": {Raw: []byte(`{"ipv6Primary": true, "addressesFromPools":[{"name":"name"}]}`)},
+			},
+			variableName:  "network",
+			expectedValue: true,
+			expectedFound: true,
+			expectedErr:   false,
+			object:        Network{},
+			expepctedVariableObject: Network{
```
Comment (killianmuldoon) [inline exp/runtime/topologymutation/variables_test.go:255]: nit
```suggestion
			expectedVariableObject: Network{
```

File: exp/runtime/topologymutation/variables_test.go [hunk 5]
```diff
@@ -130,3 +131,147 @@ func Test_GetStringTemplateVariable(t *testing.T) {
...
+
+	tests := []struct {
+		name                    string
+		variables               map[string]apiextensionsv1.JSON
+		variableName            string
+		expectedValue           bool
+		expectedFound           bool
+		expectedErr             bool
+		object                  Network
+		expepctedVariableObject interface{}
```
Comment (killianmuldoon) [inline exp/runtime/topologymutation/variables_test.go:209]: nit
```suggestion
		expectedVariableObject interface{}
```

**Review by killianmuldoon — COMMENTED** [review]
Reviewed the prod code and a little bit on the testing. Some comments on the signature of these functions to try to keep them consistent and clear.

**Review by killianmuldoon — COMMENTED** [review]
/area runtime-sdk

**Review by killianmuldoon — COMMENTED** [review]
I thiink the `TODO` on `L31` should be covered by this PR, right?

File: exp/runtime/topologymutation/variables.go (hunk 1 above)
Comment (lubronzhan) [reply exp/runtime/topologymutation/variables.go:61]: The GetVariable already implements a wrapper around not found error
```
https://github.com/kubernetes-sigs/cluster-api/blob/main/exp/runtime/topologymutation/variables.go#L33-L37
```
I'm ok to change them all to return error directly.

**Comment by lubronzhan** [comment]
`pull-cluster-api-apidiff-main ` failed because there is API change on the function

**Comment by killianmuldoon** [comment]
>pull-cluster-api-apidiff-main failed because there is API change on the function

This should be fine - as it's not required for merge. We just use it as a signal that something in the API has changed

**Comment by killianmuldoon** [comment]
/test pull-cluster-api-e2e-full-main

To run the tests that use these functions.

**Comment by lubronzhan** [comment]
Thanks. `pull-cluster-api-e2e-full-main` passed, I fixed the unit tests

**Comment by lubronzhan** [comment]
Hi @fabriziopandini could you help take a look since it's related to runtime-extension? Thanks

**Comment by sbueringer** [comment]
> Hi @sbueringer do you have time to review? Thanks

Time yes, approval no

**Comment by lubronzhan** [comment]
Hi @sbueringer is it a good time now?

File: exp/runtime/topologymutation/variables_test.go [hunk 6]
```diff
@@ -82,51 +88,197 @@ func Test_GetStringTemplateVariable(t *testing.T) {
...
+				g.Expect(patchvariables.IsNotFoundError(err)).To(BeTrue())
+			}
+		})
+	}
+}
+
+func Test_GetBoolVariable(t *testing.T) {
+	g := NewWithT(t)
+
+	varA := apiextensionsv1.JSON{Raw: toJSON("true")}
```
Comment (sbueringer) [inline exp/runtime/topologymutation/variables_test.go:146]:
```suggestion
	varA := apiextensionsv1.JSON{Raw: []byte(`true`)}
```

Bool should look like this in an apiextensionsv1.JSON (I looked at some of our unit tests in internal/topology/variables/cluster_variable_defaulting_test.go and I think I tested this when I wrote them)

File: exp/runtime/topologymutation/variables.go [hunk 7]
```diff
@@ -17,39 +17,78 @@ limitations under the License.
...
+	if err != nil {
+		return false, err
 	}
-	return stringValue, true, nil
+
+	// Unquote the JSON string.
+	stringValue, err := strconv.Unquote(string(value.Raw))
+	if err != nil {
+		return false, err
+	}
```
Comment (sbueringer) [inline exp/runtime/topologymutation/variables.go:67]:
```suggestion
```
Should not be needed for a bool

File: exp/runtime/topologymutation/variables.go [hunk 8]
```diff
@@ -17,39 +17,78 @@ limitations under the License.
...
 	patchvariables "sigs.k8s.io/cluster-api/internal/controllers/topology/cluster/patches/variables"
 )
 
-// TODO: Add func for validating received variables are of the expected types
-// TODO: Add func for converting complex variables into go structs and/or other basic types.
+// TODO: Add func for validating received variables are of the expected types.
 
 // GetVariable get the variable value.
-func GetVariable(templateVariables map[string]apiextensionsv1.JSON, variableName string) (*apiextensionsv1.JSON, bool, error) {
+func GetVariable(templateVariables map[string]apiextensionsv1.JSON, variableName string) (*apiextensionsv1.JSON, error) {
```
Comment (sbueringer) [inline exp/runtime/topologymutation/variables.go:33]: I'm generally fine with dropping the bool. Seems cleaner and idiomatic to check the resulting error, but the function that folks can use to check if the error is a not found error is from an internal package. 

Should we export / alias that one?

**Review by sbueringer — COMMENTED** [review]
Yup now is a good time. Sorry also lost track a bit of this PR.

**Comment by lubronzhan** [comment]
Almost forgot, let me get back to this tomorrow

File: exp/runtime/topologymutation/variables.go (hunk 8 above)
Comment (lubronzhan) [reply exp/runtime/topologymutation/variables.go:33]: Make sense, let me export it

Comment (lubronzhan) [reply exp/runtime/topologymutation/variables.go:33]: Then should I use this new one to replace all other previous `patchvariables.IsNotFoundError(err)`

Comment (lubronzhan) [reply exp/runtime/topologymutation/variables.go:33]: I just replaced the tests for this variable.go so that user can reference as example

Comment (sbueringer) [reply exp/runtime/topologymutation/variables.go:33]: I would say let's replace all usages (there are also a few more in variables_test.go)

**Comment by sbueringer** [comment]
Last one from my side: https://github.com/kubernetes-sigs/cluster-api/pull/9670#discussion_r1470974820

Feel free to also directly squash

File: exp/runtime/topologymutation/variables.go (hunk 8 above)
Comment (lubronzhan) [reply exp/runtime/topologymutation/variables.go:33]: Cool

**Comment by sbueringer** [comment]
Thx!

/lgtm
/approve

**Pull request merged**